### ISSUE TITLE
What should usesEveryRelationshipName do for MatchCdOd?

### ISSUE THREAD

**Issue by marcellussiegburg** [fmidue/modelling-tasks#275]
During the refactorings of #273 it appeared that `usesEveryRelationshipName` has accidentally been implemented wrong for MatchCdOd. The fact was generated for the first CD ruling out / or forcing the appearance of all relationships in that CD for all combinations of CDs.

I see two ways the parameter could be understood in this case (here for: `Just True`)
1. all ODs need to have link names for all relationship names that occur in any CD
2. all ODs need to have all link names that are possible names in the desired combination

The former is more obvious if CD1 (or CD2) does not have a relationship name that CD2 (or CD1) has. The latter is more obvious if CD1 and CD2 do not have a relationship name that CD3 has.

**Comment by jvoigtlaender** [comment]
I have no immediate hunch of "what would be better" of the two options, but maybe it is also worth considering to not even have this configuration option for the `MatchCdOd` tasks? That is, forcing it being set to `Nothing`?

I guess the option was originally introduced for the sake of the `DifferentNames` task type? And it only exists for `MatchCdOd` as well because they share the configuration type and parts of the implementation? Maybe you remember more of this, specifically whether there were Match-Task reasons/instances for wanting to have this option here.

**Comment by jvoigtlaender** [comment]
Also, just in case: The wrong implementation you observed cannot impact the intended use in DifferentNames? Doesn't it also work with "n CDs, that are then combined"? That is, the issue you observed cannot manifest in this context ("constraints meant on the first CD impost too much on the other CDs as well") like it does for MatchCdOd`?

**Comment by marcellussiegburg** [comment]
> but maybe it is also worth considering to not even have this configuration option for the MatchCdOd tasks? That is, forcing it being set to Nothing?

I think that is a good Idea and probably best (considering that there are at least two ways of interpreting it)

> The wrong implementation you observed cannot impact the intended use in DifferentNames?

Yes, it cannot.

> Doesn't it also work with "n CDs, that are then combined"?

Yes, but they all share the same names (but shuffled around).

**Comment by jvoigtlaender** [comment]
Is a similar problem lurking for `completelyInhabited`? There, the constraint (for all combinations of CDs) is based on what the `nonAbstractClassNames` are for the first CD.

**Comment by marcellussiegburg** [comment]
> Is a similar problem lurking for `completelyInhabited`?

Yes, it is! However it does not influence our scenarios at the moment (as `abstractClassNames` is always `[]` and `classNames` are the same across different CDs for our task types).

**Comment by jvoigtlaender** [comment]
I assume we will never (also in the future not) use a mutation in `MatchCdOd` that removes or adds a class. Because the fact that some class is missing would lead to too trivial matchings (or ruling out matchings) on the ODs.

But turning a concrete class diagram abstract might actually be an interesting feature in this context, since it would allow to check additional knowledge about class diagrams, and it would be non-trivial to solve such instances.

So, maybe it should at least be preserved as an issue somewhere,
- that supporting abstract classes in `MatchCdOd` might become relevant,
- that if this is done there is the equivalent of the problem found here for relationship names lurking,
- and that it would probably be addressed by ensuring that if "turn a class abstract" is contained in the allowed mutations in a `MatchCdOd` config, then the config checker should make sure that `completelyInhabited` is set to `Nothing`.

**Comment by marcellussiegburg** [comment]
> but maybe it is also worth considering to not even have this configuration option for the MatchCdOd tasks? That is, forcing it being set to Nothing?

I think the option could still be made available if `allowedCdMutations` does not contain any of `AddRelationship`, `RemoveRelationship` or `MutateRelationship ChangeKind`. Because in this case the different ways to interpret it:

> 1. all ODs need to have link names for all relationship names that occur in any CD
> 2. all ODs need to have all link names that are possible names in the desired combination

are the same

**Comment by jvoigtlaender** [comment]
Makes sense.

For the record, there might superficially be a further possible interpretation by someone reading the config:

3. for each relationship name that occurs in the CDs, at least one OD needs to have a link name

That is, interpretating the "all are used" across the collection of all ODs rather than per OD. But given that `usesEveryRelationshipName` is part of `objectConfig` and that is also otherwise specifying properties of each OD in itself, it is probably okay.
